main.js

```diff
@@ -124,20 +159,21 @@ if (config.downloadJAD) {
   loadingMIDletPromises.push(new Promise(function(resolve, reject) {
     JARStore.loadJAR("midlet.jar").then(function(loaded) {
       if (loaded) {
+        if (!document.hidden) {
+          // Show the splash screen as soon as possible.
+          showSplashScreen();
```

Comment:
One improvement we might be able to make is to front-load all the code that blocks displaying the Downloading…/Starting… screens. Currently, the code to display one or the other lives in main.js, which is the last script we load. And we load a lot of scripts before it. If we front-loaded it, we might shorten the time period between the OS splash screen (black screen with app icon) and the first app splash screen (Downloading…/Starting…).

But that would require more significant changes, like moving up the JAR store library (and its dependencies) in the load order. And this branch still makes that time period better, since it replaces the background screen (white screen with app icon), which currently appears there, with a solid black screen that is a less jarring transition.

Another option would be to identify the install run via some other mechanism than the absence of the midlet JAR from the JAR store. For example, we could use the *lastVersion* property that the compiled method cache sets in local storage. Then we could pick between Downloading…/Starting… earlier on that basis. This would work most of the time, except when users fail to download the midlet and then kill/restart the app. And we could make it work all the time by setting another property in local storage that definitely identified the presence of the midlet JAR.

Anyway, these are both worth considering, but I'd do any additional work here in a followup.